### ISSUE TITLE
Access Violation on x64

### ISSUE THREAD

**Issue by radischevo** [dotnet/coreclr#23516]
Some time ago we experienced a series of weird access violation exceptions in an ASP.NET web application running on .NET 4.7.2. We found a `NullReferenceException` in our code, and the fix made the problem disappear. 

After some investigation we were able to provide a [minimal demo](https://github.com/radischevo/RyuJitBugDemo) that can always be reproduced on .NET 4.7.2 and .NET Core 2.1.x (both Windows and MacOS) when built in Release configuration against x64. 

```
public static class Program
{
    public static void Main(string[] args)
    {
        Container container = null;
        Console.WriteLine($"{container.Item}.");
    }
}

public struct Option
{
    private object _foo;
    private int _bar;
}

public class Container
{
    public Option Item { get; }
}
```

The exception details are:

```Problem Event Name:	APPCRASH
  Application Name:	NullReferenceExceptionDemo.exe
  Application Version:	0.0.0.0
  Application Timestamp:	aa7bfd6c
  Fault Module Name:	clr.dll
  Fault Module Version:	4.7.3062.0
  Fault Module Timestamp:	5ab9567c
  Exception Code:	c0000005
  Exception Offset:	0000000000003f60
  OS Version:	6.1.7601.2.1.0.256.48
  Locale ID:	1049
  Additional Information 1:	dd8a
  Additional Information 2:	dd8acd90d29aab5b23f9fa4eb1fb1c16
  Additional Information 3:	571d
  Additional Information 4:	571d8a78cbdeb9ddd7155af6ae2ca7fd
```

Please note, that the issue is tighly related to the fact, that `Option` is a value type that has at least two fields: one of a reference type, and the other one of arbitrary type. If the structure contains one field, or both of the fields are value types, the problem does not occur. 

As the same example runs as expected on .NET 4.7.2 when built against AnyCPU platform, I suspect this to be a bug in RyuJIT. 

I was not able to reproduce this on .NET Core 2.2 and 3.0.

**Comment by AndyAyersMS** [comment]
This is the same issue as in #19444. An AV in `CORINFO_HELP_ASSIGN_BYREF` is not translated to the appropriate managed exception.

Fix was done in 3.0 and ported to 2.2. Port to 2.1 (see #20440) was put on hold pending more understanding of impact. 

@radischevo can you say more about your experience debugging and working around this issue, and the impact it had on your use of .Net?

**Comment by onyxmaster** [comment]
@AndyAyersMS It was me who dealt with this issue in the first place. It was pretty bad, since the problem manifested itself when one of the external services we use (a search engine) started returning different results. 
Shortly: all our production backends (10 IIS instances) crashed with ExecutionEngineException almost simultaneosly. The attempts to correlate the problem with any code changes by rolling back failed, since the code was there all the time, it's the change of the input data that triggered the different code path, also rolling back 3 days worth of commits and redeploying didn't help.

It was pretty obvious that this is either memory corruption (but disabling all `unsafe` components we had didn't help), netstandard<->.NET Framework incompatibility (this was my theory for most of the time, see below), or JIT/Runtime issue.
Unfortunately, the stack traces I obtained (both managed and native) weren't too helpful, so I just abandoned the debugging and isolated it to a specific component manually, differentiating on "what could break", disabling it altogether.

In the end it was about 3 hours of downtime and some more gray hair for my beard =)

I was quite busy the following two days since I was visiting a conference, so I had no chance to take a look at it, but I assumed it was [this](http://blog.rytmis.net/2018/05/03/executionengineexception-in-asp-net-mvc-when-mixing-net-framework-and-net-standard-assemblies/), since we had some related issues last year. 
Fortunately I was wrong (I _really_ would hate to go over System.Web.Http + netstandard2 libs + .NET Framework runtime again), since @radischevo did a great job of isolating the issue itself to a minimal case.

P.S. We don't care that much about .NET Core 2.1, whatever we can run on .NET Core is being ran on latest released version (2.2), it's the .NET Framework 4.7.2 which our backends is running on, so our priority would be fixing this in the full one.

**Comment by AndyAyersMS** [comment]
Thanks for this bug report, and for the details above.

As far as getting this fixed in .Net Framework, it would be helpful if you could also report this via https://developercommunity.visualstudio.com/, if you haven't done so already.

cc @RussKeldorph

**Comment by radischevo** [comment]
Well, my part of the job was relatively small. After @onyxmaster got the app working by disabling all requests to external services, I had only to find the failing one enabling them one-by-one. 
Then I ran the application (built In Debug) on a dev machine, and immediately got a `NullReferenceException`. That time I could not believe the null dereference was the root cause of the problem, so I just fixed it and then examined different code paths, async stuff, `XmlSerializer` creation, etc. But after several hours of debugging I realized that the only thing that made sense was the fix made in the beginning.
Shame on me, there were no tests for this module ;)

I've [reported the bug](https://developercommunity.visualstudio.com/content/problem/509540/access-violation-on-net-framework-472-x64-only.html) to Developer Community, hope the fix for 4.7.2 will be released soon.
